src/backend/renderer/mod.rs

```diff
@@ -0,0 +1,173 @@
+use std::error::Error;
+
+use cgmath::{prelude::*, Matrix3, Vector2};
+#[cfg(feature = "wayland_frontend")]
+use wayland_server::protocol::{wl_shm, wl_buffer};
+
+use crate::backend::SwapBuffersError;
+#[cfg(feature = "renderer_gl")]
+pub mod gles2;
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Transform {
+    Normal,
+    _90,
+    _180,
+    _270,
+    Flipped,
+    Flipped90,
+    Flipped180,
+    Flipped270,
+}
+
+impl Transform {
+    pub fn matrix(&self) -> Matrix3<f32> {
+        match self {
+	        Transform::Normal => Matrix3::new(
+                1.0, 0.0, 0.0,
+                0.0, 1.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::_90 => Matrix3::new(
+                0.0, -1.0, 0.0,
+                1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::_180 => Matrix3::new(
+                -1.0, 0.0, 0.0,
+                0.0, -1.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::_270 => Matrix3::new(
+                0.0, 1.0, 0.0,
+                -1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::Flipped => Matrix3::new(
+                -1.0, 0.0, 0.0,
+                0.0, 1.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::Flipped90 => Matrix3::new(
+                0.0, 1.0, 0.0,
+                1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::Flipped180 => Matrix3::new(
+                1.0, 0.0, 0.0,
+                0.0, -1.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+            Transform::Flipped270 => Matrix3::new(
+                0.0, -1.0, 0.0,
+                -1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0,
+            ),
+        }
+    }
+
+    pub fn invert(&self) -> Transform {
+        match self {
+            Transform::Normal => Transform::Normal,
+            Transform::Flipped => Transform::Flipped,
+            Transform::_90 => Transform::_270,
+            Transform::_180 => Transform::_180,
+            Transform::_270 => Transform::_90,
+            Transform::Flipped90 => Transform::Flipped270,
+            Transform::Flipped180 => Transform::Flipped180,
+            Transform::Flipped270 => Transform::Flipped90,
+        }
+    }
+
+    pub fn transform_size(&self, width: u32, height: u32) -> (u32, u32) {
+        if *self == Transform::_90
+            || *self == Transform::_270
+            || *self == Transform::Flipped90
+            || *self == Transform::Flipped270
+        {
+            (height, width)
+        } else {
+            (width, height)
+        }
+    }
+}
+
+#[cfg(feature = "wayland-frontend")]
+impl From<wayland_server::protocol::wl_output::Transform> for Transform {
```

Comment:
I suspect we'll likely need the conversion the other way around as well.

Reply:
At some point, yes. ^^